src/gui/plugins/scene3d/Scene3D.cc

```diff
@@ -2307,6 +2548,7 @@ void RenderWindowItem::keyReleaseEvent(QKeyEvent *_e)
       _e->accept();
     }
     this->DeselectAllEntities(true);
+    // TODO(john): Delete model preview on Escape press
```

Comment:
I had to force focus in hover event for escape to work:

```
diff --git a/src/gui/plugins/scene3d/Scene3D.cc b/src/gui/plugins/scene3d/Scene3D.cc
index 31e2849a..5e0e23c2 100644
--- a/src/gui/plugins/scene3d/Scene3D.cc
+++ b/src/gui/plugins/scene3d/Scene3D.cc
@@ -2463,6 +2463,7 @@ void RenderWindowItem::hoverMoveEvent(QHoverEvent *_e)
   if (pos == oldPos)
     return;
   this->dataPtr->renderThread->ignRenderer.NewHoverEvent(pos);
+  this->forceActiveFocus();
 }
 
 /////////////////////////////////////////////////
@@ -2549,6 +2550,7 @@ void RenderWindowItem::keyReleaseEvent(QKeyEvent *_e)
     }
     this->DeselectAllEntities(true);
     // TODO(john): Delete model preview on Escape press
+    this->dataPtr->renderThread->ignRenderer.TerminatePreviewModel();
   }
 }
```

I added some printout and it looks like the` OnFocusWindow` function is not triggered any more in this branch?

another issue that may or may not be related: the context menu for the render window now pops up in the top left corner on right click instead of on the object clicked

Reply:
These problems are both related in some way: They both begin to occur when the RenderWindowItem class starts accepting hover events with `this->setAcceptHoverEvents(true);` 

I'm really not sure why beginning to accept hover events would affect these other functionalities.  There isn't any documentation that I've read so far that shows accepting hover events should interfere with other functionalities.  I'm not sure how we might want to move forward given these conflicting functionalities.  The hover event is needed for the shapes plugin, but breaking already working features is not something that should happen.  Possibly more research is needed.  I've tried a few things including making the render window a sibling of the mouse area in the `GzScene3D.qml` file, which seemed to perhaps be a step in the right direction, but it also introduced novel issues to the problem while solving some others.

I can take a further look soon and hopefully try to find some sort of solution.

Reply:
I think I've solved the issue.  After some investigation, I came to a conclusion that was supported by some of my testing.  The reason as to why introducing the hover events in the `RenderWindow` breaks standing functionalities is due to the nature of how Qt handles events.  There isn't quite much documentation for the HoverEvents, but I have made the assumption that their rules for behavior are the same as that of the KeyEvents, and [here](https://doc.qt.io/qt-5/qtquick-input-focus.html), rule 4 states that if a key event is accepted, propagation will stop.  This is the key to this problem.  In `GzScene3D.qml`, the `MouseArea` has `hoverEnabled: true`, meaning it will accept all of the hover events.  This carries the implication that there will be no propagation to other entities, in this case RenderWindow.  

Likewise, if we have `RenderWindow` accepting the hover events, and disable the `MouseArea`'s `hoverEnabled` attribute, the `MouseArea` will no longer provide the force focus capability and will also no longer provide `mouseX` and `mouseY` values, which render window uses to set the location of an opened entity context menu - this explains why the menus began to appear at the top corner of the window once the mouse area's hover capabilities were disabled.

In the end, I decided to handle the hover position through the `MouseArea`.  I added a callback setter function that will update the hovered mouse position when a new position is detected within the mouse area, rather than having the `RenderWindow` accept the hover events directly.  In this way, we are capturing these hover events in a single place to hopefully avoid any conflicts goings forward.